miden-lib/asm/sat/internal/faucet.masm

```diff
@@ -0,0 +1,271 @@
+use.std::collections::smt
+
+use.miden::sat::internal::account
+use.miden::sat::internal::asset
+use.miden::sat::internal::asset_vault
+use.miden::sat::internal::layout
+
+# FUNGIBLE ASSETS
+# ==================================================================================================
+
+#! Mints a fungible asset associated with the fungible faucet the transaction is being executed
+#! against.
+#!
+#! Panics:
+#! - If the transaction is not being executed against a fungible faucet.
+#! - If the fungible asset being minted is not associated with the faucet the transaction is
+#!   being executed against.
+#! - If the asset is not well formed.
+#! - If the total issuance after minting is greater than the maximum amount allowed.
+#!
+#! Stack: [ASSET]
+#! Outputs: [ASSET]
+#!
+#! - amount is the amount of the fungible asset to mint.
+#! - ASSET is the asset that was minted.
+export.mint_fungible_asset
+    # assert that we are executing a transaction against a fungible faucet (access checks)
+    exec.account::get_id dup exec.account::is_fungible_faucet assert
+    # => [fungible_faucet_id, ASSET]
+
+    # assert that the asset is associated with the faucet the transaction is being executed against
+    dup.1 assert_eq
+    # => [ASSET]
+
+    # assert the asset is well formed
+    exec.asset::validate_fungible_asset
+    # => [ASSET]
```

Comment:
I wonder if there is some redundancy in these checks. I think it should be enough to verify that:
1. The asset is a well-formed fungible asset.
2. Faucet ID of the asset matches the current account ID.

So, maybe something like this would be enough?:
```
export.mint_fungible_asset
    # assert the asset is a well-formed fungible asset
    exec.asset::validate_fungible_asset
    # => [ASSET]

   # assert that we are executing a transaction against the faucet of this asset
   dup exec.account::get_id assert_eq
    # => [ASSET]
```

Reply:
Another thought: maybe it would make sense to create a procedure for validating asset ownership. Maybe something like `asset::validate_fungible_asset_ownership` with the following interface:
```
inputs:  [account_id, ASSET, ...]
outputs: [ASSET, ...]
```

We could have something similar for non-fungible assets as well. This way, all logic which makes assumptions about structure of `ASSET` is encapsulated in the `asset.masm` module.

Reply:
I like this idea of introducing `validat_*_asset_ownership` abstraction procedures, however I'm not sure if `ownership` is the correct term to use here. `ownership` implies that the asset is owned by the provided `facuet_id` however this is not necessarily the case.  The asset could in fact be owned by a different user account or a note.  Maybe we could have `validate_*_asset_faucet`.

Reply:
Agreed that `ownership` is probably the wrong word here. Another option is to use `origin` - e.g., `validate_fungible_asset_origin`.